lib/cadet/courses/courses.ex

```diff
@@ -7,8 +7,124 @@ defmodule Cadet.Courses do
+
+        {:ok, Map.put_new(course, :assessment_types, assessment_types)}
+    end
+  end
+
+  @doc """
+  Updates the general course configuration for the specified course
+  """
+  @spec update_course_config(integer, %{}) ::
+          {:ok, %Course{}} | {:error, {:bad_request, String.t()} | {:error, Ecto.Changeset.t()}}
+  def update_course_config(course_id, params) when is_ecto_id(course_id) do
+    case retrieve_course(course_id) do
+      nil ->
+        {:error, {:bad_request, "Invalid course id"}}
+
+      course ->
+        course
+        |> Course.changeset(params)
+        |> Repo.update()
+    end
+  end
+
+  defp retrieve_course(course_id) when is_ecto_id(course_id) do
+    Course
+    |> where(id: ^course_id)
+    |> Repo.one()
+  end
+
+  @doc """
+  Updates the assessment configuration for the specified course
+  """
+  @spec update_assessment_config(integer, integer, integer, integer, integer) ::
+          {:ok, %AssessmentConfig{}} | {:error, Ecto.Changeset.t()}
+  def update_assessment_config(course_id, order, early_xp, hours_before_decay, decay_rate) do
+    AssessmentConfig
+    |> join(:inner, [ac], at in AssessmentType, on: at.order == ^order)
+    |> where([ac, at], at.course_id == ^course_id)
+    |> Repo.one()
+    |> AssessmentConfig.changeset(%{
+      early_submission_xp: early_xp,
+      hours_before_early_xp_decay: hours_before_decay,
+      decay_rate_points_per_hour: decay_rate
+    })
+    |> Repo.update()
+  end
+
+  @doc """
+  Updates the Assessment Types for the specified course
+  """
+  @spec update_assessment_types(integer, list()) :: :ok | {:error, {:bad_request, String.t()}}
+  def update_assessment_types(course_id, params) do
+    if not is_list(params) do
+      {:error, {:bad_request, "Invalid parameter(s)"}}
+    else
+      params_length = params |> length()
+
+      with true <- params_length <= 5,
+           true <- params_length >= 1,
+           true <- params |> Enum.reduce(true, fn elem, acc -> acc and is_binary(elem) end),
+           true <-
+             params |> Enum.map(fn elem -> String.capitalize(elem) end) |> Enum.uniq() |> length() ===
+               params_length do
+        (params ++ List.duplicate(nil, 5 - params_length))
+        |> Enum.with_index(1)
+        |> Enum.each(fn {elem, idx} ->
+          case elem do
+            nil ->
+              AssessmentType
+              |> where(course_id: ^course_id)
+              |> where(order: ^idx)
+              |> Repo.delete_all()
+
+            _ ->
+              AssessmentType.changeset(%AssessmentType{}, %{
+                course_id: course_id,
+                order: idx,
+                type: elem
+              })
+              |> Repo.insert(
+                on_conflict: {:replace, [:type]},
+                conflict_target: [:course_id, :order]
+              )
+          end
+        end)
+      else
+        false ->
+          {:error, {:bad_request, "Invalid parameter(s)"}}
+      end
+    end
+  end
```

Comment:
This function should be split into 3 or 4 functions:

1. insert type
2. update type
3. delete type
4. change order of all types

1 & 2 might be combined into an upsert if you prefer

Reply:
Thanks for the comment! I have resolved a few comments. I have updated config routes, combine the assessment config table and refactor the update assessment config flow.